### PR TITLE
[Executor] Include tokens in span attributes for streaming mode

### PR DESCRIPTION
**Pull request by lumoslnt**
# Description

For streaming mode, the output is a generator, and we cannot gather information from it, so we need to extract the information we need after the output has been consumed and update to a new span, which named Iterated(<original_span_name>). In this PR, we gather LLM model and calculate the tokens based on the consumed output and update these to span attributes.

## Key Changes:

Package Installation:
* [`.github/workflows/build_doc_ci.yml`](diffhunk://#diff-3ad70f8240b8dcfe2acb849af506bedab9a120ca0ed5efd0134857bf2b17af33R39-R45): Added installation of `promptflow-tracing` package.
* [`.github/workflows/promptflow-executor-e2e-test.yml`](diffhunk://#diff-3c2611e02a5661ec57a2d45f6f9b50cb83c30e460cfef589c6f9bdcdea3bbc25R107-R108): Added installation of `promptflow-tracing` package.
* [`.github/workflows/promptflow-executor-unit-test.yml`](diffhunk://#diff-8ba5006c2b0ff4c2980241cb8e2bfa339a8424ef06d9547f7398946eb9ed7c74R112-R113): Added installation of `promptflow-tracing` package.
* [`.github/workflows/promptflow-global-config-test.yml`](diffhunk://#diff-c2d48285feb43cb0dfa958473cc67ee514eab8140a17fc45535713d1697febb7R63-R68): Added installation of `promptflow-tracing` package.
* [`.github/workflows/promptflow-sdk-cli-test.yml`](diffhunk://#diff-94bdef7b4df710f08c32e0e3a7a27a14f093e487c5605b894b887d9b65a64dc7R81-R82): Added installation of `promptflow-tracing` package.

Codebase Adjustments:
* [`src/promptflow-tracing/promptflow/tracing/_openai_utils.py`](diffhunk://#diff-88d8c0365919464ea99a0089fbc3d4b9959e1736d5e42d03a288d3d29b0d66d5R2-L5): Several changes were made to this file, including the removal of the `CalculatingMetricsError` class, the addition of the `OpenAIResponseParser` abstract base class and its subclasses, and modifications to several methods. [[1]](diffhunk://#diff-88d8c0365919464ea99a0089fbc3d4b9959e1736d5e42d03a288d3d29b0d66d5R2-L5) [[2]](diffhunk://#diff-88d8c0365919464ea99a0089fbc3d4b9959e1736d5e42d03a288d3d29b0d66d5R41) [[3]](diffhunk://#diff-88d8c0365919464ea99a0089fbc3d4b9959e1736d5e42d03a288d3d29b0d66d5L63-R76) [[4]](diffhunk://#diff-88d8c0365919464ea99a0089fbc3d4b9959e1736d5e42d03a288d3d29b0d66d5L85-R99) [[5]](diffhunk://#diff-88d8c0365919464ea99a0089fbc3d4b9959e1736d5e42d03a288d3d29b0d66d5L127-R129) [[6]](diffhunk://#diff-88d8c0365919464ea99a0089fbc3d4b9959e1736d5e42d03a288d3d29b0d66d5L154-L157) [[7]](diffhunk://#diff-88d8c0365919464ea99a0089fbc3d4b9959e1736d5e42d03a288d3d29b0d66d5L204-R263)
* [`src/promptflow-tracing/promptflow/tracing/_trace.py`](diffhunk://#diff-580941184737186a94e3e9c06e467e86c06ce846d30ffa42c1c43b45812ddcdcR20): Several changes were made to this file, including the addition of the `collect_openai_tokens_for_streaming` method, modifications to the `enrich_span_with_trace_type` method, and the addition of the `enrich_span_with_llm_output` method. [[1]](diffhunk://#diff-580941184737186a94e3e9c06e467e86c06ce846d30ffa42c1c43b45812ddcdcR20) [[2]](diffhunk://#diff-580941184737186a94e3e9c06e467e86c06ce846d30ffa42c1c43b45812ddcdcR46-R55) [[3]](diffhunk://#diff-580941184737186a94e3e9c06e467e86c06ce846d30ffa42c1c43b45812ddcdcR133-L176) [[4]](diffhunk://#diff-580941184737186a94e3e9c06e467e86c06ce846d30ffa42c1c43b45812ddcdcL224-R257)

Other Changes:
* [`.github/workflows/promptflow-tracing-unit-test.yml`](diffhunk://#diff-e0a27b9b9fb9a60d557556e51d8fea5b13d57448b4a4565358e200db66c2fe58L95-R95): Lowered the thresholds for the unit test.
* [`scripts/check_enforcer/check_enforcer.py`](diffhunk://#diff-13d98bbe46293159f6c3381c310213d6c2749796ec32784b4626a371d4ca0260L50-R54): Commented out the `sdk_cli_global_config_tests` dictionary.

# All Promptflow Contribution checklist:
- [x] **The pull request does not introduce [breaking changes].**
- [ ] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [x] **I have read the [contribution guidelines](../CONTRIBUTING.md).**
- [ ] **Create an issue and link to the pull request to get dedicated review from promptflow team. Learn more: [suggested workflow](../CONTRIBUTING.md#suggested-workflow).**

## General Guidelines and Best Practices
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### Testing Guidelines
- [x] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: src/promptflow/tests/executor/e2etests/test_traces.py [hunk 1]
```diff
@@ -384,26 +384,29 @@ def assert_otel_traces_with_prompt(self, dev_connections, flow_file, inputs, pro
...
+        "flow_file, inputs, is_stream, expected_span_length",
         [
-            ("openai_chat_api_flow", get_chat_input(False), 3),
-            ("openai_completion_api_flow", get_comletion_input(False), 3),
-            ("llm_tool", {"topic": "Hello", "stream": False}, 4),
-            ("flow_with_async_llm_tasks", get_flow_sample_inputs("flow_with_async_llm_tasks"), 6),
+            ("openai_chat_api_flow", get_chat_input(False), False, 3),
+            ("openai_chat_api_flow", get_chat_input(True), True, 4),
+            ("openai_completion_api_flow", get_comletion_input(False), False, 3),
+            ("openai_completion_api_flow", get_comletion_input(True), True, 4),
```
Comment (thy09) [inline src/promptflow/tests/executor/e2etests/test_traces.py:392]: get_completion_input?

File: src/promptflow/tests/executor/e2etests/test_traces.py [hunk 2]
```diff
@@ -531,13 +534,13 @@ def validate_span_list(self, span_list, line_run_id, expected_span_length):
...
 
-    def validate_openai_tokens(self, span_list):
+    def validate_openai_tokens(self, span_list, is_stream):
         span_dict = {span.context.span_id: span for span in span_list}
         expected_tokens = {}
         for span in span_list:
             tokens = None
             # Validate the openai tokens are correctly set in the llm trace.
-            if span.attributes.get("function", "") in LLM_FUNCTION_NAMES:
+            if self._is_llm_function_name(span.attributes.get("function", ""), is_stream):
```
Comment (thy09) [inline src/promptflow/tests/executor/e2etests/test_traces.py:543]: Add some comments for the logic

File: src/promptflow/tests/executor/e2etests/test_traces.py (hunk 1 above)
Comment (lumoslnt) [reply src/promptflow/tests/executor/e2etests/test_traces.py]: Fixed.

File: src/promptflow/tests/executor/e2etests/test_traces.py (hunk 2 above)
Comment (lumoslnt) [reply src/promptflow/tests/executor/e2etests/test_traces.py:543]: Added,

File: src/promptflow/promptflow/_core/tracer.py [hunk 3]
```diff
@@ -297,60 +308,70 @@ def enrich_span_with_input(span, input):
...
     ) as span:
-        span.set_attributes(original_span.attributes)
+        enrich_span_with_original_attributes(span, original_span.attributes)
         generator_proxy = GeneratorProxy(generator)
         yield from generator_proxy
         generator_output = generator_proxy.items
 
-        # Enrich LLM span for openai steaming message
-        # TODO: Enrich LLM token count for streaming scenario
+        # Enrich LLM span for OpenAI steaming output
```
Comment (thy09) [inline src/promptflow/promptflow/_core/tracer.py:338]: streaming output

Comment (lumoslnt) [reply src/promptflow/promptflow/_core/tracer.py]: Fixed.

**Review by thy09 — APPROVED** [review]

File: src/promptflow/promptflow/tracing/_trace.py [hunk 4]
```diff
@@ -120,60 +131,70 @@ def enrich_span_with_input(span, input):
...
+def enrich_span_with_original_attributes(span, attributes):
+    try:
+        span.set_attributes(attributes)
+    except Exception as e:
+        logging.warning(f"Failed to enrich span with original attributes: {e}")
+
+
+def enrich_span_with_llm(span, model, generated_message):
+    try:
+        span.set_attribute("llm.model", model)
```
Comment (zzn2) [inline src/promptflow/promptflow/tracing/_trace.py:185]: We are going to remove this attribute to `llm.response.model`, see #2270

File: src/promptflow/promptflow/tracing/_trace.py [hunk 5]
```diff
@@ -120,60 +131,70 @@ def enrich_span_with_input(span, input):
...
-            if chunks:
-                text = "".join(chunks)
-                message = {"content": text, "role": role}
-                span.set_attribute("llm.generated_message", serialize_attribute(message))
-        serialized_output = serialize_attribute(generator_output)
-        span.set_attribute("output", serialized_output)
+            from openai.types.completion import Completion
+
+            if generator_output and isinstance(generator_output[0], (ChatCompletionChunk, Completion)):
+                parser = OpenAIResponseParser(generator_output)
```
Comment (zzn2) [inline src/promptflow/promptflow/tracing/_trace.py:167]: (Optional) Let's separate OpenAIResponseParser into two, one for Chat, and another for Completion. They can inherit from the same abstract parent class.

Comment (lumoslnt) [reply src/promptflow/promptflow/tracing/_trace.py]: Done.

File: .github/actions/step_sdk_setup/action.yml [hunk 6]
```diff
@@ -28,6 +28,17 @@ runs:
       echo "########### pip freeze ###########"
       pip freeze
     working-directory: ${{ inputs.scriptPath }}
+  - uses: snok/install-poetry@v1
```
Comment (crazygao) [inline .github/actions/step_sdk_setup/action.yml:31]: move this step to install python.

File: .github/actions/step_sdk_setup/action.yml [hunk 7]
```diff
@@ -28,6 +28,17 @@ runs:
...
+  - name: 'Build and install: promptflow-tracing'
+    shell: pwsh
+    run: |
+      Set-PSDebug -Trace 2
+      poetry build
```
Comment (crazygao) [inline .github/actions/step_sdk_setup/action.yml:36]: Just poetry install will do the whole work.

File: .github/workflows/build_doc_ci.yml [hunk 8]
```diff
@@ -18,7 +18,32 @@ env:
...
+        scriptPath: ${{ env.testWorkingDirectory }}
+    - name: Upload Wheel
+      if: always()
+      uses: actions/upload-artifact@v3
+      with:
+        name: wheel
+        path: |
+          ${{ github.workspace }}/src/promptflow/dist/*.whl
+          ${{ github.workspace }}/src/promptflow-tools/dist/*.whl
+          ${{ github.workspace }}/src/promptflow-tracing/dist/*.whl
```
Comment (crazygao) [inline .github/workflows/build_doc_ci.yml:43]: why this step?
pip install -e will do the work for you.

**Review by crazygao — APPROVED** [review]

**Review by thy09 — DISMISSED** [review]

File: .github/actions/step_sdk_setup/action.yml (hunk 6 above)
Comment (lumoslnt) [reply .github/actions/step_sdk_setup/action.yml]: Removed.

File: .github/actions/step_sdk_setup/action.yml (hunk 7 above)
Comment (lumoslnt) [reply .github/actions/step_sdk_setup/action.yml]: Refined.

File: .github/workflows/build_doc_ci.yml (hunk 8 above)
Comment (lumoslnt) [reply .github/workflows/build_doc_ci.yml]: Refined.

**Review by crazygao — APPROVED** [review]

**Review by thy09 — APPROVED** [review]

**Pull request merged**